### PR TITLE
drivers: dma: add DMA driver for NXP PXP engine

### LINKED ISSUE #59921: Enabling PXP accelearion for lvgl for i. MXRT1064
Greetings all,
Currently it isn't possible to use pxp gpu acceleration for 2D lvgl rendering in ZephyrOS. Pxp acceleration not only is necessary for faster rendering, but also for display rotation, which doesn't seem to be possible (with acceptable performance) without the pxp accel.

Here is some research of files that would be need. But just adding them to a Project isn't enough: 
- lv_gpu/lv_gpu_nxp.c, lv_gpu/lv_gpu_nxp.h: low level drawing calls for LVGL
- lv_gpu/lv_gpu_nxp_osa.c, lv_gpu/lv_gpu_osa.h: default implementation of OS-specific functions (bare metal and FreeRTOS only)
     - optional, required only if `LV_USE_GPU_NXP_PXP_AUTO_INIT` is set to 1
- PXP related code depends on two drivers provided by MCU SDK. These drivers need to be added to project:
    - fsl_pxp.c, fsl_pxp.h: PXP driver
    - fsl_cache.c, fsl_cache.h: CPU cache handling functions

### PR DESCRIPTION
**Pull request by danieldegrasse**
The NXP Pixel Pipeline Engine (PXP) is a 2D DMA engine capable of
accelerating display rotation, color space conversion, and limited
2D blending operations. This DMA driver only supports rotation of a
framebuffer, via a set of custom dma_slot values.

The driver enables rotation in concert with the ELCDIF driver. For example, a user can configure the ELCDIF driver to leverage the PXP to rotate the display by 270 degrees with the following settings:

```
CONFIG_DMA=y
CONFIG_MCUX_ELCDIF_PXP=y
CONFIG_MCUX_ELCDIF_PXP_ROTATE_270=y
```

If used, LVGL can then be configured to rotate the display using the following call (LVGL rotates the display counter clockwise, and the PXP rotates it clockwise):

```
lv_disp_set_rotation(lv_disp_get_default(), LV_DISP_ROT_90);
```

Fixes #59921

### CODE REVIEW

**Comment by danieldegrasse** [comment]
@Gigate, will this cover your use case? I'm not sure if there is much more of the PXP functionality we can expose via the DMA API, but I have verified that display rotation functions (with acceptable performance) on the RT1060 and RT1170 EVKs using LVGL, provided rotation is configured as I outline in this PR.

**Comment by Gigate** [comment]
@danieldegrasse thanks for the quick response. I just checked out out this PR, however this doesn't work for my mimxrt1064_evk board. When enabling the configs the screen isn't rotated but the display is blinking in a 0.5 sec interval. Also rotating by 180 degrees doesn't work either, but then there is no blinking.

File: include/zephyr/drivers/dma/dma_mcux_pxp.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_DMA_MCUX_PXP_H_
+#define ZEPHYR_INCLUDE_DRIVERS_DMA_MCUX_PXP_H_
+
+#define DMA_MCUX_PXP_CMD_MASK (0xE0)
+#define DMA_MCUX_PXP_CMD_SHIFT (0x5)
+
+#define DMA_MCUX_PXP_FMT_MASK (0x1F)
+#define DMA_MCUX_PXP_FMT_SHIFT (0x0)
```
Comment (jfischer-no) [inline include/zephyr/drivers/dma/dma_mcux_pxp.h:14]: Please do not add unnecessary parentheses.

**Comment by danieldegrasse** [comment]
> @danieldegrasse thanks for the quick response. I just checked out out this PR, however this doesn't work for my mimxrt1064_evk board. When enabling the configs the screen isn't rotated but the display is blinking in a 0.5 sec interval. Also rotating by 180 degrees doesn't work either, but then there is no blinking.

Sorry, I should have noticed you were using an RT1064 EVK. This PR requires that LVGL (or whatever graphics middleware is in use) calls `display_write` with a display buffer equal in size to the screen (or the screen rotated by 90 degrees, when the PXP is enabled).

The relevant Kconfigs are configured on the RT1170 EVK and RT1060 EVK. You can look here to see the settings you need: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/mimxrt1060_evk/Kconfig.defconfig#L60. In particular, you should set:

```
CONFIG_LV_Z_VDB_SIZE=100
CONFIG_LV_Z_DOUBLE_VDB=y
CONFIG_LV_Z_FULL_REFRESH=y
CONFIG_LV_Z_VDB_ALIGN=32
CONFIG_LV_Z_BITS_PER_PIXEL=16 (assuming you use an RGB565 display)
```

File: include/zephyr/drivers/dma/dma_mcux_pxp.h (hunk 1 above)
Comment (danieldegrasse) [reply include/zephyr/drivers/dma/dma_mcux_pxp.h:14]: Noted, fixed in latest push

**Review by jfischer-no — APPROVED** [review]

**Comment by Gigate** [comment]
@danieldegrasse thanks, I got it working now. However I also got sw rotation working under v3.2.0 in the meantime. Comparing both solutions I noticed, that the sw rotation is way more performant (measuring with CONFIG_LV_USE_PERF_MONITOR=y) . PXP rotation 8% CPU usage and sw rotation 2%.  Also when enabling the debug log under the PXP rotation I get spammed with

- mpu: mpu_configure_region: Configure MPU region at index 0x3
- mpu: region_init: [3] 0x8011ae00 0x150b000a
- mpu: region_allocate_and_init: Program MPU region at index 0x3

Is there something else I need to configure?

**Comment by danieldegrasse** [comment]
> Also when enabling the debug log under the PXP rotation I get spammed with

I see these logs with or without PXP rotation enabled. They are related to MPU regions being dynamically configured when task switching occurs.

> PXP rotation 8% CPU usage and sw rotation 2%. Also when enabling the debug log under the PXP rotation

I'm currently testing with the latest version of Zephyr, and LVGL 8.3.7 (the version of the module integrated with Zephyr). I'm not able to reproduce these results with software rotation. I assume you are setting `sw_rotate=1`? I see poor performance, and CPU utilization bouncing between 7 and 20%. With the PXP enabled I see similar CPU usage (maybe a bit more consistent, around 15%), but the subjective performance of UI elements is much better.

**Comment by Gigate** [comment]
> I see these logs with or without PXP rotation enabled. They are related to MPU regions being dynamically configured when task switching occurs.

Ok same for me, my sw rotated version was under a different zephyr version (v3.2.0). There the overall performance is way better, but this most likely is a different issue?!

> I'm currently testing with the latest version of Zephyr, and LVGL 8.3.7 (the version of the module integrated with Zephyr). I'm not able to reproduce these results with software rotation. I assume you are setting `sw_rotate=1`? I see poor performance, and CPU utilization bouncing between 7 and 20%. With the PXP enabled I see similar CPU usage (maybe a bit more consistent, around 15%), but the subjective performance of UI elements is much better.

I checked out this PR, and updated to LVGL 8.3.7 (I had 8.2.0 before). Comparing sw rotation and pxp rotation on the mimxrt1064_evk still yields inconclusive results. CPU Performance is bouncing for my sample app in both cases between 4% and 9%. However with CONFIG_LV_DISP_DEF_REFR_PERIOD=16 (to get a higher framerate) the FPS counter is at ~ 42 FPS with PXP rotation and at ~62 with sw rotation.

**Comment by danieldegrasse** [comment]
> I checked out this PR, and updated to LVGL 8.3.7 (I had 8.2.0 before). Comparing sw rotation and pxp rotation on the mimxrt1064_evk still yields inconclusive results. CPU Performance is bouncing for my sample app in both cases between 4% and 9%. However with CONFIG_LV_DISP_DEF_REFR_PERIOD=16 (to get a higher framerate) the FPS counter is at ~ 42 FPS with PXP rotation and at ~62 with sw rotation.

I am running the LVGL widgets sample on the RT1060. With the PXP performing rotation, I see FPS bouncing between ~40 and ~20. CPU utilization bounces between ~13 and ~27 percent.

With software rotation, I see FPS more consistently around 62, and CPU utilization bounces between ~13 and ~9 percent. However, the moment I actually interact with the sample (by swiping between screens), FPS drops to 4, and CPU utilization spikes. I do not see these performance dips with the PXP enabled

My current guess for the reasons behind this performance are the following:
- When the PXP is enabled, `CONFIG_LV_Z_FULL_REFRESH` must be on. This means even a minor display update requires updating the entire framebuffer. Since the PXP is involved, the framebuffer rotation will always occur. I suspect this is why the idle CPU usage remains high, because even partial updates to the display require a full framebuffer refresh
- When the PXP is disabled, `CONFIG_LV_Z_FULL_REFRESH` is off. The software rotation is more efficient for partial display updates, as only a subset of the display frame will be modified. However once the full display needs to be updated (IE when swiping between screens in the widgets view), the software rotation becomes less efficient. This is when the FPS dips

To summarize, I'm not entirely sure why the PXP yields high CPU utilization as I am not deeply familiar with the LVGL performance counter code. However, I see subjectively smoother performance when using the PXP, and I do not see framerate drops when interacting with the sample.

**Comment by kisvegabor** [comment]
I cc @nicusorcitu from NXP. Probably he can tell  more about it.

**Comment by nicusorcitu** [comment]
Hi,

The LVGL CPU load represent the usage of the CPU during execution of the timer handlers. It is not the overall CPU system usage.

1. CONFIG_LV_DISP_DEF_REFR_PERIOD shall be set according to the display refresh rate.
When you set the CONFIG_LV_DISP_DEF_REFR_PERIOD to 33 ms it means the timer callback to render a frame will get called at every 33ms (it will try to satisfy the 30FPS you expect). During the timer refresh execution process, the PXP might take over and do some of the rendering operations (like blitting images, or filling areas - simple rectangles). The CPU load DOES NOT TAKE in consideration the PXP work or not during rendering process.
With CPU only, if rendering a frame takes 16ms, the CPU load might look like 50% - this is an estimation because there might be other callback timers to get executed and CPU load measure them all.
With PXP enabled, if rendering a frame takes 20ms, the CPU load might look like 60% - when in fact this is not true because part of the frame could have been handled by PXP hardware while the CPU was basically in the idle state.

On FreeRTOS I did a fix for that, so that the CPU load calculation is based on the real time spend in IdleTask. Hopefully we can do a better way of calculation CPU usage in lvgl v9 based on the info provided by the OS. @kisvegabor 

2. You shall not decrease the value of CONFIG_LV_DISP_DEF_REFR_PERIOD to 16ms unless you really expect 60FPS. Doing that, the timer callback will get executed every 16ms and will consume useless the resources if your screen is only 30Hz.

3. The screen rotation process is outside of the rendering process. Screen rotation time is not measured in the CPU load calculation.
@kisvegabor this is also another request we had it to support the rotation as a task in lvgl v9 so that we can measure that one too. Or split the rotation between multiple units (CPU, PXP, VGLite...) if available.

Thanks,
Nicu.

**Comment by kisvegabor** [comment]
> @kisvegabor this is also another request we had it to support the rotation as a task in lvgl v9 so that we can measure that one too. Or split the rotation between multiple units (CPU, PXP, VGLite...) if available.

I've already added it to the [ROADMAP](https://github.com/lvgl/lvgl/commit/3b5731e40bffab2666f7cb9f0f09669f1c972ad9) of v9.

**Review by yvanderv — APPROVED** [review]
Looks good to me

**Review by decsny — APPROVED** [review]
I think this is fine for now to enable use of the PXP
